src/Compilers/CSharp/Portable/Lowering/ExtensionMethodBodyRewriter.cs

```diff
@@ -0,0 +1,160 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Roslyn.Utilities;
+using ReferenceEqualityComparer = Roslyn.Utilities.ReferenceEqualityComparer;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// Rewrites method body lowered in context of an extension method to a body
+    /// that corresponds to its implementation form.
+    /// </summary>
+    internal sealed class ExtensionMethodBodyRewriter : BoundTreeToDifferentEnclosingContextRewriter
+    {
+        private readonly SourceExtensionImplementationMethodSymbol _implementationMethod;
+
+        /// <summary>
+        /// Maps parameters and local functions from original enclosing context to corresponding rewritten symbols for rewritten context.
+        /// </summary>
+        private ImmutableDictionary<Symbol, Symbol> _symbolMap;
+
+        private RewrittenMethodSymbol _rewrittenContainingMethod;
+
+        public ExtensionMethodBodyRewriter(MethodSymbol sourceMethod, SourceExtensionImplementationMethodSymbol implementationMethod)
+        {
+            Debug.Assert(sourceMethod is not null);
+            Debug.Assert(implementationMethod is not null);
+            Debug.Assert(MethodCompiler.TryGetCorrespondingExtensionImplementationMethod(sourceMethod) == (object)implementationMethod);
+
+            _implementationMethod = implementationMethod;
+            _symbolMap = ImmutableDictionary<Symbol, Symbol>.Empty.WithComparers(ReferenceEqualityComparer.Instance, ReferenceEqualityComparer.Instance);
+
+            bool haveExtraParameter = sourceMethod.ParameterCount != implementationMethod.ParameterCount;
+            if (haveExtraParameter)
+            {
```

Comment:
Consider asserting that `implementationMethod.ParameterCount - 1 == sourceMethod.ParameterCount`